### ISSUE TITLE
Regarding submitit checkpointing

### ISSUE THREAD

**Issue by netw0rkf10w** [facebookresearch/detr#237]
Hello,

In `run_with_submitit.py`, you set an `output_dir` [for each job](https://github.com/facebookresearch/detr/blob/5e66b4cd15b2b182da347103dd16578d28b49d69/run_with_submitit.py#L70), thus the checkpoint path for a job is `[checkpoint_dir]/[job_id]/checkpoint.pth`. Now for example, if the current job 245 has reached the time limit and has been killed (while training hasn't finished yet), the next job 246 will be launched, but it won't resume from `[checkpoint_dir]/245/checkpoint.pth` because it will only look for `[checkpoint_dir]/246/checkpoint.pth`, which doesn't exist.

Am I correct?

Thanks in advance for your answer!

**Comment by fmassa** [comment]
Hi,

I believe [submitit uses `requeue` on timeout](https://github.com/facebookincubator/submitit/blob/89fbe3144fa5ad3674178ecba8fd53da21f072ca/submitit/slurm/slurm.py#L193), so the jobs would be requeued with the same job id, so this shouldn't be an issue. @jrapin can correct me if I'm wrong.

Did you face any issues as what you mentioned when using submitit and facing a timeout ? I don't recall experiencing those issues myself.

**Comment by netw0rkf10w** [comment]
@fmassa Thanks for the reply. You are right, the job is re-queued with the same ID. I've just managed to get it work, and it works as expected.

One limitation of the current script is that some computation may be wasted if the timeout is reached in the middle of training. For example, suppose that training has been running for 1 hour since the last checkpoint, and it is about to save a new checkpoint, but then it is stopped because of the timeout, which means the 1 hour of computation has been wasted.

I have added some timeout control to `main.py` that, before starting a new epoch, estimates if that epoch will surpass the given time limit or not (based on the average time of the last few epochs), if it will then stop. With this timeout control, one can submit a sequence of jobs: job `n` can start only after job `n-1` has finished (i.e. the `afterok` dependency of Slurm). No computation is wasted using this approach, but then it doesn't seem obvious how to use it with `submitit`...

**Comment by fmassa** [comment]
I believe this should be possible, at least by mixing submitit code within the training script.
Otherwise, in the launch script, you could probably have the script wait on the job finishing in order to launch a new job.
Something like [what submitit has in the documentation](https://github.com/facebookincubator/submitit/tree/master#an-example-is-worth-a-thousand-words-performing-an-addition)

It might require a few iterations to get the whole thing right, but I think it should be doable.

**Comment by jrapin** [comment]
> I believe [submitit uses `requeue` on timeout](https://github.com/facebookincubator/submitit/blob/89fbe3144fa5ad3674178ecba8fd53da21f072ca/submitit/slurm/slurm.py#L193), so the jobs would be requeued with the same job id, so this shouldn't be an issue. @jrapin can correct me if I'm wrong.

Hey, sorry for answering a bit late, was busy on a deadline :s
Yes, it is requeued, see here for the actual code
https://github.com/facebookincubator/submitit/blob/89fbe3144fa5ad3674178ecba8fd53da21f072ca/submitit/slurm/slurm.py#L191-L194

How often does the job checkpoints ? If it does every minute or so, and it starts back from the last checkpoint, it's probably sufficient. Otherwise, it's possible to configure what happens at checkpoint time with the checkpoint method of a callable:
https://github.com/facebookincubator/submitit/blob/master/docs/checkpointing.md
Hope that helps!

**Comment by netw0rkf10w** [comment]
@fmassa @jrapin Thanks a lot for your replies!

> How often does the job checkpoints ? If it does every minute or so, and it starts back from the last checkpoint, it's probably sufficient.

I'm not sure I understand your question correctly. DETR's training saves a checkpoint every epoch (which takes ~30 minutes for the base model).

> Otherwise, it's possible to configure what happens at checkpoint time with the checkpoint method of a callable:
> https://github.com/facebookincubator/submitit/blob/master/docs/checkpointing.md
> Hope that helps!

Thanks. According to the documentation, it seems that the `checkpoint()` function is called whenever Slurm sends a timeout signal:

```python
def checkpoint(self, checkpointpath: str) -> submitit.helpers.DelayedSubmission:
        # the checkpoint method is called asynchroneously when the slurm manager
        # sends a preemption signal, with the same arguments as the __call__ method
```

This means that the waste of computation that I described in the previous message cannot happen: even if the time limit is reached in the middle of a training, the current checkpoint is saved and will be resumed after the job has been requeued.

**Comment by jrapin** [comment]
> Thanks. According to the documentation, it seems that the checkpoint() function is called whenever Slurm sends a timeout signal. This means that the waste of computation that I described in the previous message cannot happen: even if the time limit is reached in the middle of a training, the current checkpoint is saved and will be resumed after the job has been requeued.

Exactly. However, I just checked the code, and the checkpoint method in detr actually does not save the current state (in other words, this checkpoint method does not checkpoint :D), it will only requeue a job that is starting from the last checkpoint:

https://github.com/facebookresearch/detr/blob/699bf53f3e3ecd4f000007b8473eda6a08a8bed6/run_with_submitit.py#L52-L63

I do not know if that would be a good practice or not to checkpoint from the middle of an epoch, it's not for submitit to decide in any case ;)

I forgot one thing for context, actually by default submitit does not requeue on timeouts, only on preemptions. It only requeues on timeouts if there is a checkpoint method. This is to avoid requeueing jobs that are bound to timeout again (some cause, same consequence). And it only requeues a number of times (`slurm_max_num_timeouts` in the executor init)

**Comment by netw0rkf10w** [comment]
Many many thanks to @jrapin for the explanations! And I am so sorry for the late reply, I have been taken and overwhelmed lately by some deadline.

> I forgot one thing for context, actually by default submitit does not requeue on timeouts, only on preemptions.

Maybe one can somehow fake a preemption signal from Python to force a re-queue:

```python
for epoch in range(args.start_epoch, args.epochs):
    ...
    train_one_epoch()
    evaluate()
    # check for job timeout
    if estimated_time_next_epoch > args.time_limit:
        # better stop here and send a Slurm preemption signal so that the job will be requeued
        send_preemption_signal()
        break
```

Let me do some research to see if this is possible.

**Comment by jrapin** [comment]
> Maybe one can somehow fake a preemption signal from Python to force a re-queue

There is no need, maybe I should have emphasized that it does not requeue **by default**, i.e. is the `checkpoint` method is not implemented. Here the method is implemented, so it should be requeued on both timeouts and preemptions
